### PR TITLE
Feature/requirement lists ng

### PR DESCRIPTION
**Pull request by a-w50**
Refactored, edited and cleaned changes from wip/requirement_lists

### CODE REVIEW

**Comment by tmolitor-stud-tu** [comment]
why squashing this into only two commits?
by keeping the smaller commits its really easy to revert one of the changes (for example the "don'T allow default values thing), when squasing everything into one or two commits, undoing individual changes is much more complicated.

that's a policy thing and thus that's a question for all
@hikinggrass 
@a-w50 
@caller 
@corneliusclaussen 
@pionix-juraed

**Comment by corneliusclaussen** [comment]
We decided to have the main history contain larger blocks instead of fine grained commits

File: everestjs/everestjs.cpp [hunk 1]
```diff
@@ -381,93 +380,99 @@ static Napi::Value boot_module(const Napi::CallbackInfo& info) {
...
+            // if this was a requirement with min_connections == 1 and max_connections == 1,
+            // this will be simply a single connection, but an array of connections otherwise
+            // (this array can have only one entry, if only one connection was provided, though)
+            const bool is_list = req_route_list.is_array();
+            if (!is_list) {
+                req_route_list = json::array({req_route_list});
             }
+            auto req_mod_vars_array = Napi::Array::New(env);
+            auto req_mod_cmds_array = Napi::Array::New(env);
+            for (uint64_t i = 0; i < req_route_list.size(); i++) {
```
Comment (hikinggrass) [inline everestjs/everestjs.cpp]: I changed this to uint64_t because we use this type for the Requirement index

File: include/utils/types.hpp [hunk 2]
```diff
@@ -26,19 +26,35 @@ using ValueCallback = std::function<void(Value)>;
...
 
+struct ModuleInfo {
+    std::string name;
+    std::vector<std::string> authors;
+    std::string license;
+    std::string id;
+};
+
+struct Requirement {
+    Requirement(const std::string& requirement_id, uint64_t index) : id(requirement_id), index(index){};
```
Comment (hikinggrass) [inline include/utils/types.hpp]: Changed index to uint64_t to prevent narrowing conversion

File: include/utils/types.hpp [hunk 3]
```diff
@@ -26,19 +26,35 @@ using ValueCallback = std::function<void(Value)>;
...
+        if (id < rhs.id) {
+            return true;
+        } else if (id == rhs.id) {
+            return (index < rhs.index);
+        } else {
+            return false;
+        }
+    }
+    const std::string id;
+    uint64_t index;
```
Comment (hikinggrass) [inline include/utils/types.hpp]: Changed index to uint64_t to prevent narrowing conversion

**Review by hikinggrass — APPROVED** [review]
I had to change the index of Requirement to uint64_t to prevent narrowing conversion, or was that a conscious decision  to use int here? Then we have to reflect that in the ev-cli code generator as well.

File: everestjs/everestjs.cpp (hunk 1 above)
Comment (tmolitor-stud-tu) [reply everestjs/everestjs.cpp]: No, no concious decisision, uint64_t is fine

**Comment by tmolitor-stud-tu** [comment]
> We decided to have the main history contain larger blocks instead of fine grained commits

I think thats a bad decision for various reasons (bisect, document what change has which effect etc.)

**Pull request merged**